### PR TITLE
MAINT: Use userpointers to avoid NPY_MAXARGS in iternext()

### PR DESCRIPTION
**Pull request by seberg**
Let's get rid of all uses of NPY_MAXARGS in the iterator (and most other places).

I would prefer not to do dynamic allocations here, so re-using the userpointers.
Unfortunately, that needs a bit of care in the buffer copy code to get right.

---

I paused for a bit on whether I forgot to check for core-offset (I didn't and even if, changing the iterator position manually doesn't use the `prev_dataptrs` mechanism).  So, the test tests something that cannot be broken for multiple reasons, but after having it, thought we might as well add it.

### CODE REVIEW

File: numpy/_core/src/multiarray/nditer_api.c [hunk 1]
```diff
@@ -2157,19 +2157,27 @@ npyiter_copy_to_buffers(NpyIter *iter, char **prev_dataptrs)
...
+        /* NOTE: we re-use the NIT_USERPTRS for this check (in iternext) */
+        assert(prev_dataptrs == NULL || prev_dataptrs == NIT_USERPTRS(iter));
+        /*
+         * The prev_dataptrs must match to re-use buffers.  Also coreoffset
+         * must not be 0 to be able to re-use it.
+         * (As of writing coreoffset != 0 implies prev_dataptrs == NULL.)
+         */
+        if (prev_dataptrs == NULL || prev_dataptrs[iop] != dataptrs[iop]) {
+            NIT_OPITFLAGS(iter)[iop] &= ~NPY_OP_ITFLAG_BUF_REUSABLE;
         }
```
Comment (seberg) [inline numpy/_core/src/multiarray/nditer_api.c:2170]: Admittedly, this now unsets the flag on writeable arguments as well, which is just unnecessary :thinking:.
(Something I'll add to the iterator issue: buffer re-use for writeable ops could be interesting.)

File: numpy/_core/src/multiarray/nditer_api.c [hunk 2]
```diff
@@ -2187,9 +2196,8 @@ npyiter_copy_to_buffers(NpyIter *iter, char **prev_dataptrs)
             NPY_IT_DBG_PRINT("    marking operand %d for buffer reuse\n", iop);
             NIT_OPITFLAGS(iter)[iop] |= NPY_OP_ITFLAG_BUF_REUSABLE;
         }
-        else if (prev_dataptrs == NULL || prev_dataptrs[iop] != dataptrs[iop] ||
-                 bufferdata->coreoffset) {
```
Comment (seberg) [inline numpy/_core/src/multiarray/nditer_api.c:2191]: Seemed not useful to check.  We just always unset the flag (even when it isn't set).

**Review by mhvk — APPROVED** [review]
Yes, this makes sense. I was worried a bit whether the user pointers are exposed at all, but they are still always reset to those of the buffers anyway, so it should be good.

**Comment by seberg** [comment]
N.B. for you @mhvk: I changed the setup slightly in another PR.  `USERPTRS` was `NBF_PTRS` before (for buffered iteration).
`USERPTRS` is now the name of the pointers the user is given, when the user needs a copy of them (buffer or external loop, without both mutating the pointers was always invalid).

As for enabling an arbitrary number of operands.  For the C-API this is now easy, there is only the one use left in the new buffer-setup function.

The tedious part is that we also need to fix the Python exposure.  That isn't terribly hard, but that code can use a bit of love...
Maybe including a `.tp_vectorcall` to make places where we abuse it for broadcasting.

---

Otherwise will just put it in a bit since there is an approval, thanks for the review! :).

**Comment by mhvk** [comment]
Super, sounds good, let's get it in.

**Pull request merged**